File: backend/services/product_service.py
from fastapi import HTTPException
from models.product import Product
from schemas.product import ProductCreate, ProductUpdate
from sqlalchemy.orm import Session


class ProductService:
    def __init__(self, db: Session):
        self._db = db

    def get_all_products(self):
        return self._db.query(Product).all()

    def get_product_by_id(self, product_id: int):
        product = self._db.query(Product).filter(Product.id == product_id).first()
        if not product:
            raise HTTPException(status_code=404, detail="Product not found")
        return product

    def delete_product(self, product_id: int):
        product = self.get_product_by_id(product_id)
        self._db.delete(product)
        self._db.commit()
        return {"message": "Product deleted successfully"}

    def create_product(self, product_data: ProductCreate):
        existing_product = (
            self._db.query(Product)
            .filter(Product.sku_code == product_data.sku_code)
            .first()
        )
        if existing_product:
            raise HTTPException(status_code=400, detail="SKU code already exists")

        if product_data.quantity < 0:
            raise HTTPException(status_code=400, detail="Quantity must be non-negative")

        if product_data.price < 0:
            raise HTTPException(status_code=400, detail="Price must be non-negative")

        product = Product(**product_data.model_dump())
        self._db.add(product)
        self._db.commit()
        self._db.refresh(product)
        return product

    def update_product(self, product_id: int, product_data: ProductUpdate):
        product = self.get_product_by_id(product_id)
        update_data = product_data.model_dump(exclude_none=True)

        if "sku_code" in update_data:
            existing = (
                self._db.query(Product)
                .filter(
                    Product.sku_code == update_data["sku_code"],
                    Product.id != product_id,
                )
                .first()
            )
            if existing:
                raise HTTPException(status_code=400, detail="SKU code already exists")

        if "quantity" in update_data and update_data["quantity"] < 0:
            raise HTTPException(status_code=400, detail="Quantity must be non-negative")

        if "price" in update_data and update_data["price"] < 0:
            raise HTTPException(status_code=400, detail="Price must be non-negative")

        for key, value in update_data.items():
            setattr(product, key, value)

        self._db.commit()
        self._db.refresh(product)
        return product

File: frontend/services/product_service.ts
import api from "./api"
import type {
  ProductCreate,
  ProductDeleteResponse,
  ProductListResponse,
  ProductResponse,
  ProductUpdate,
} from "@/types/product"

const PRODUCT_API_PATH = "product"

const getAllProducts = async (): Promise<ProductListResponse> => {
  const response = await api.get<ProductListResponse>(`${PRODUCT_API_PATH}/`)
  return response.data
}

const getProductById = async (productId: number): Promise<ProductResponse> => {
  const response = await api.get<ProductResponse>(`${PRODUCT_API_PATH}/${productId}`)
  return response.data
}

const createProduct = async (productData: ProductCreate): Promise<ProductResponse> => {
  const response = await api.post<ProductResponse>(
    `${PRODUCT_API_PATH}/create`,
    productData,
  )
  return response.data
}

const updateProduct = async (
  productId: number,
  productData: ProductUpdate,
): Promise<ProductResponse> => {
  const response = await api.put<ProductResponse>(
    `${PRODUCT_API_PATH}/${productId}`,
    productData,
  )
  return response.data
}

const deleteProduct = async (productId: number): Promise<ProductDeleteResponse> => {
  const response = await api.delete<ProductDeleteResponse>(
    `${PRODUCT_API_PATH}/${productId}`,
  )
  return response.data
}

export const productService = {
  getAllProducts,
  getProductById,
  createProduct,
  updateProduct,
  deleteProduct,
}
